src/common/substrait/src/df_expr.rs

```diff
@@ -0,0 +1,731 @@
+use substrait_proto::protobuf::expression::reference_segment::{
+    ReferenceType as SegReferenceType, StructField,
+};
+use substrait_proto::protobuf::expression::{
+    FieldReference, ReferenceSegment, RexType, ScalarFunction,
+};
+use substrait_proto::protobuf::function_argument::ArgType;
+use substrait_proto::protobuf::Expression;
+
+use crate::context::ConvertorContext;
+use crate::error::{
+    EmptyExprSnafu, InvalidParametersSnafu, MissingFieldSnafu, Result, UnsupportedExprSnafu,
+};
+
+/// Convert substrait's `Expression` to DataFusion's `Expr`.
+pub fn to_df_expr(ctx: &ConvertorContext, expression: Expression, schema: &Schema) -> Result<Expr> {
+    let expr_rex_type = expression.rex_type.context(EmptyExprSnafu)?;
+    match expr_rex_type {
+        RexType::Literal(_) => UnsupportedExprSnafu {
+            name: "substrait Literal expression",
+        }
+        .fail()?,
+        RexType::Selection(selection) => convert_selection_rex(*selection, schema),
+        RexType::ScalarFunction(scalar_fn) => convert_scalar_function(ctx, scalar_fn, schema),
+        RexType::WindowFunction(_)
+        | RexType::IfThen(_)
+        | RexType::SwitchExpression(_)
+        | RexType::SingularOrList(_)
+        | RexType::MultiOrList(_)
+        | RexType::Cast(_)
+        | RexType::Subquery(_)
+        | RexType::Enum(_) => UnsupportedExprSnafu {
+            name: format!("substrait expression {:?}", expr_rex_type),
+        }
+        .fail()?,
+    }
+}
+
+/// Convert Substrait's `FieldReference` - `DirectReference` - `StructField` to Datafusion's
+/// `Column` expr.
+pub fn convert_selection_rex(selection: FieldReference, schema: &Schema) -> Result<Expr> {
+    if let Some(FieldReferenceType::DirectReference(direct_ref)) = selection.reference_type
+    && let Some(SegReferenceType::StructField(field))  = direct_ref.reference_type{
+        let column_name = schema.column_name_by_index(field.field as _).to_string();
+         Ok(Expr::Column(Column{ relation: None, name: column_name }))
+    } else {
+        InvalidParametersSnafu{reason:"Only support direct struct reference in Selection Rex"}.fail()
+    }
+}
+
+pub fn convert_scalar_function(
+    ctx: &ConvertorContext,
+    scalar_fn: ScalarFunction,
+    schema: &Schema,
+) -> Result<Expr> {
+    // convert argument
+    let mut inputs = VecDeque::with_capacity(scalar_fn.arguments.len());
+    for arg in scalar_fn.arguments {
+        if let Some(ArgType::Value(sub_expr)) = arg.arg_type {
+            inputs.push_back(to_df_expr(ctx, sub_expr, schema)?);
+        } else {
+            InvalidParametersSnafu {
+                reason: "Only value expression arg is supported to be function argument",
+            }
+            .fail()?;
+        }
+    }
+
+    // convert this scalar function
+    // map function name
+    let anchor = scalar_fn.function_reference;
+    let fn_name = ctx.find_scalar_fn(anchor).context(InvalidParametersSnafu {
```

Comment:
```suggestion
    let fn_name = ctx.find_scalar_fn(anchor).with_context(|| InvalidParametersSnafu {
```

Reply:
why this is better

Reply:
Using context will construct the InvalidParametersSnafu, thus formatting and allocating a new string at each call.

Reply:
Similar to the relationship between unwrap_or and unwrap_or_else

Reply:
I see lots (hundreds) of this idiom occurrences in the code base. Maybe you should open an issue or write a clippy rule to constrain this further.